qualtran/bloqs/mcmt/bloq_with_specialized_single_qubit_control_test.py

```diff
@@ -0,0 +1,171 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from typing import Optional, Sequence, Tuple
+from unittest.mock import ANY
+
+import attrs
+import pytest
+from attrs import evolve, frozen
+
+from qualtran import AddControlledT, Bloq, CtrlSpec, Signature
+from qualtran.bloqs.mcmt import And
+from qualtran.bloqs.mcmt.bloq_with_specialized_single_qubit_control import (
+    get_ctrl_system_for_bloq_with_specialized_single_qubit_control,
+)
+from qualtran.resource_counting import CostKey, GateCounts, get_cost_value, QECGatesCost
+
+
+@frozen
+class AtomWithSpecializedControl(Bloq):
+    cv: Optional[int] = None
+
+    @property
+    def signature(self) -> 'Signature':
+        n_ctrl = 1 if self.cv is not None else 0
+        return Signature.build(ctrl=n_ctrl, q=2)
+
+    def get_ctrl_system(self, ctrl_spec: 'CtrlSpec') -> Tuple['Bloq', 'AddControlledT']:
+        return get_ctrl_system_for_bloq_with_specialized_single_qubit_control(
+            self,
+            ctrl_spec,
+            current_control_bit=self.cv,
+            bloq_with_ctrl=attrs.evolve(self, cv=1),
+            ctrl_reg_name='ctrl',
+            bloq_without_ctrl=attrs.evolve(self, cv=None),
+            bloq_with_ctrl_0=attrs.evolve(self, cv=0),
```

Comment:
I guess one complaint would be that this makes a couple copies of the bloq, when we really only need one(-ish?) of them. The attrs.evolve calls could be seen as redundant as well. What if we have a base function that accepts a callable and then two variants of the interface that provide a friendlier(?) interface to the callable version

```python
def get_special_1bit_ctrl(unctrl_bloq, get_ctrl_bloq: Callable[]):
  ctrl_bloq = get_ctrl_bloq(ctrl_spec.get_single_ctrl_bit())

def get_attrs_special_1bit_ctrl(unctrl_bloq, cv_attr_name: str = 'cv')
  return get_special_1bit_ctrl(unctrl_bloq, lambda x: attrs.evolve(unctrl_bloq, **{cv_attr_name: x}))

def get_special_1bit_ctrl_from_list(unctrl_bloq, bloq_with_1_ctrl, bloq_with_0_ctrl):
  return get_special_1bit_ctrl(unctrl_bloq, lambda x: {0: bloq_with_1_ctrl, 1: bloq_with_0_ctrl, None: unctrl_bloq}[x])
```

Reply:
This makes sense. The main method now takes a callable, and I added a helper (third case) to accept the bloq variants.

The second case seems like more work than it is worth to support, as the control register name may depend on the bloq etc. I've updated the examples to use the new helper (third case) instead.